### PR TITLE
AG-12599 - maintainColumnOrder extended and modified

### PR DESCRIPTION
**Pull request by AndrewGlazier**

### CODE REVIEW

File: community-modules/core/src/entities/gridOptions.ts [hunk 1]
```diff
@@ -335,9 +335,9 @@ export interface GridOptions<TData = any> {
     };
     /**
      * Keeps the order of Columns maintained after new Column Definitions are updated.
-     * @default false
+     * @default 'pivotResultColumns'
      */
-    maintainColumnOrder?: boolean;
+    maintainColumnOrder?: boolean | 'primaryAndPivotResultColumns' | 'primaryColumns' | 'pivotResultColumns' | 'none';
```
Comment (peterjrreynolds) [inline community-modules/core/src/entities/gridOptions.ts:340]: Can change `primaryAndPivotResultColumns` to `all`

Also, add deprecation note on boolean to the comment here

File: documentation/ag-grid-docs/src/content/docs/column-updating-definitions/index.mdoc [hunk 2]
```diff
@@ -132,7 +132,7 @@ In other words, to have the grid correctly match Columns make sure each Column h
 The example below demonstrates the different matching strategies. Note the following:
 
 * All columns, with the exception of Country, are matched correctly. This means any column width, sort etc
-  will be kept between changes to the columns. Country will have its state reset, as it will be treated as a new column
+  will be kept between changes to the columns. Country will have its state reset, §as it will be treated as a new column
```
Comment (peterjrreynolds) [inline documentation/ag-grid-docs/src/content/docs/column-updating-definitions/index.mdoc:135]: Typo

File: community-modules/core/src/columns/columnModel.ts [hunk 3]
```diff
@@ -237,7 +231,22 @@ export class ColumnModel extends BeanStub implements NamedBean {
...
+        if (maintainColumnOrder === true) {
+            maintainColumnOrder = 'primaryAndPivotResultColumns';
+        } else if (maintainColumnOrder === false) {
+            maintainColumnOrder = 'pivotResultColumns';
+        }
+
+        const shouldSortNewColDefs =
+            maintainColumnOrder !== 'none' &&
+            (maintainColumnOrder === 'primaryAndPivotResultColumns' ||
+                maintainColumnOrder === (this.showingPivotResult ? 'pivotResultColumns' : 'primaryColumns'));
```
Comment (StephenCooper) [inline community-modules/core/src/columns/columnModel.ts:246]: Can we have this in a grid options utils function.

File: community-modules/core/src/entities/gridOptions.ts (hunk 1 above)
Comment (AndrewGlazier) [reply community-modules/core/src/entities/gridOptions.ts]: done - deprecation note done without @deprecated due to lack of deprecation of entire property

File: documentation/ag-grid-docs/src/content/docs/column-updating-definitions/index.mdoc (hunk 2 above)
Comment (AndrewGlazier) [reply documentation/ag-grid-docs/src/content/docs/column-updating-definitions/index.mdoc]: done

File: community-modules/core/src/columns/columnModel.ts (hunk 3 above)
Comment (AndrewGlazier) [reply community-modules/core/src/columns/columnModel.ts]: done

File: community-modules/core/src/entities/gridOptions.ts [hunk 4]
```diff
@@ -335,9 +335,11 @@ export interface GridOptions<TData = any> {
     };
     /**
      * Keeps the order of Columns maintained after new Column Definitions are updated.
-     * @default false
+     *
+     * The use of boolean values with `maintainColumnOrder` have been deprecated as of v32.2.0
+     * @default 'pivotResultColumns'
      */
-    maintainColumnOrder?: boolean;
+    maintainColumnOrder?: boolean | 'all' | 'primaryColumns' | 'pivotResultColumns' | 'none';
```
Comment (peterjrreynolds) [inline community-modules/core/src/entities/gridOptions.ts:342]: Should we export a type for the string values?

Comment (AndrewGlazier) [reply community-modules/core/src/entities/gridOptions.ts:342]: I don't think so - we don't for other string unions. Lots of precedent to not, and not much (if any) to do so

**Pull request merged**